module/zfs/dmu_recv.c

```diff
@@ -1302,9 +1305,11 @@ dmu_recv_begin(const char *tofs, const char *tosnap,
 	 * upper limit. Systems with less than 1GB of RAM will see a lower
 	 * limit from `arc_all_memory() / 4`.
 	 */
-	if (payloadlen > (MIN((1U << 28), arc_all_memory() / 4)))
+	if (payloadlen > (MIN((1U << 28), arc_all_memory() / 4))) {
+		crfree(cr);
+		drc->drc_cred = NULL;
 		return (E2BIG);
-
+	}
```

Comment:
While here, could add `SET_ERROR()`.

Reply:
Fair enough. I'm not a fan of unrelated changes but this one is smol and meh :)